File: algorithms/graph/search_algorithms/bfs.py
from collections import deque

# Breadth First Search algorithm:
# Time Complexity:
# O(V+E), где:
# V — количество вершин,
# E — количество рёбер.
#
# Space Complexity:
# Memory: O(V)

def bfs_on_graph(graph, start):
    n = len(graph)
    visited = [False] * n
    distances = [-1] * n

    queue = deque()
    queue.append(start)
    visited[start] = True
    distances[start] = 0

    while queue:
        node = queue.popleft()

        for nei in graph[node]:
            if not visited[nei]:
                visited[nei] = True
                distances[nei] = distances[node] + 1
                queue.append(nei)

    return distances

def bfs_on_grid(grid, start_x, start_y):
    m, n = len(grid), len(grid[0])
    distances = [[-1] * m for _ in range(n)]

    queue = deque()
    queue.append((start_x, start_y))
    distances[start_x][start_y] = 0

    directions = [(0, 1), (1, 0), (0, -1), (-1, 0)]

    while queue:
        x, y = queue.popleft()

        for dx, dy in directions:
            new_x, new_y = x + dx, y + dy

            if 0 <= new_x < n and 0 <= new_y < m:
                if grid[new_x][new_y] != "#" and distances[new_x][new_y] == -1:
                    distances[new_x][new_y] = distances[x][y] + 1
                    queue.append((new_x, new_y))

    return distances

# Where to use:
# Если ты видишь:
#
# “минимальное число шагов”
# “за сколько ходов”
# “распространение”
# “grid”
# “каждый шаг = 1 стоимость”

File: cpp_algorithms/graph/search_algorithms/bfs.cpp
#include <bits/stdc++.h>
using namespace std;
using ll = long long;
#define pb push_back

// Breadth First Search Algorithm:
//
// Time Complexity:
// O(V + E), где:
// V — количество вершин,
// E — количество рёбер.
//
// Space Complexity:
// Memory: O(V)

vector<int> bfs_on_graph(const vector<vector<int>>& graph, int start) {
    int n = graph.size();

    vector<bool> visited(n, false);
    vector<int> distances(n, -1);

    queue<int> q;

    q.push(start);
    visited[start] = true;
    distances[start] = 0;

    while (!q.empty()) {
        int node = q.front();
        q.pop();

        for (int nei : graph[node]) {
            if (!visited[nei]) {
                visited[nei] = true;
                distances[nei] = distances[node] + 1;

                q.push(nei);
            }
        }
    }

    return distances;
}


// BFS on Grid:
//
// Time Complexity:
// O(n * m)
//
// Space Complexity:
// O(n * m)

vector<vector<int>> bfs_on_grid(
    const vector<string>& grid,
    int start_x,
    int start_y
) {
    int n = grid.size();
    int m = grid[0].size();

    vector<vector<int>> distances(
        n,
        vector<int>(m, -1)
    );

    queue<pair<int, int>> q;

    q.push({start_x, start_y});
    distances[start_x][start_y] = 0;

    vector<pair<int, int>> directions = {
        {0, 1},
        {1, 0},
        {0, -1},
        {-1, 0}
    };

    while (!q.empty()) {
        auto [x, y] = q.front();
        q.pop();

        for (auto [dx, dy] : directions) {
            int new_x = x + dx;
            int new_y = y + dy;

            if (
                new_x >= 0 && new_x < n &&
                new_y >= 0 && new_y < m
            ) {
                if (
                    grid[new_x][new_y] != '#' &&
                    distances[new_x][new_y] == -1
                ) {
                    distances[new_x][new_y] =
                        distances[x][y] + 1;

                    q.push({new_x, new_y});
                }
            }
        }
    }

    return distances;
}

// Where to use:
//
// Если ты видишь:
//
// "минимальное число шагов"
// "за сколько ходов"
// "распространение"
// "grid"
// "каждый шаг = 1 стоимость"
